### PR TITLE
containerd: infer MTU from host's network interface

### LINKED ISSUE #6599: Switching to containerd results in TLS handshake timeout
## Summary

Switching from guardian to containerd results in get registry-image-resource requests showing an error
`ERRO[0010] download failed: initialize transport: Get "https://replaced.com/v2/": net/http: TLS handshake timeout 
`

## Steps to reproduce

- name: node-image
  type: registry-image-git
  icon: docker
  source:
    repository: replaced.com/node/node
    username: {{user}}
    password: {{pw}}

jobs:
- name: build-and-push
  plan:
  - get: node-image
    trigger: true
    params:
      format: oci

## Triaging info

* Concourse version: 7.0.0 
* Browser (if applicable):
* Did this used to work? When using guardian runtime yes

### LINKED ISSUE #6613: [containerd] Worker container MTU not respected
## Summary

When using the containerd runtime, the containers spawned for jobs do not have the same MTU as the concourse worker container. This means that if the worker has an MTU of lower than 1500 the containers will incorrectly perform TCP MSS, leading to randomly dropped packets and stalled connections.

## Steps to reproduce

1. Perform the standard concourse "getting started" setup with the following modifications to `docker-compose.yml`:

Add the following network to set the worker container's MTU to 1480:
```yaml
networks:                                
  default:                               
    driver: bridge                       
    driver_opts:                         
      com.docker.network.driver.mtu: 1480
```
Set the `CONCOURSE_RUNTIME` environment variable (on the worker) to `containerd`

2. Apply the following pipeline (this is an example I've found is nearly guaranteed to run into the issue within a few runs, but to be clear the issue effects essentially any pipeline that accesses the internet):

```yaml
---
resources:
  - name: debian
    type: registry-image
    icon: docker
    check_every: never
    source:
      repository: docker.io/debian
      tag: buster
jobs:
  - name: mtu-example
    plan:
      - get: debian
      - task: example
        image: debian
        config:
          platform: linux
          run:
            path: /bin/sh
            args:
              - -exc
              - |
                apt-get update -y
                apt-get install -y --no-install-recommends sl
```
3. Unpause the pipeline and start the build. It may take a few builds, but `apt` should fail to update eventually. A more precise way of detecting the issue is shown below in Expected/Actual results.

Example of one of my failures testing this:
```
E: Failed to fetch http://deb.debian.org/debian/pool/main/n/ncurses/libncurses6_6.1+20181013-2+deb10u2_amd64.deb  Connection failed [IP: 151.101.6.132 80]
E: Unable to fetch some archives, maybe run apt-get update or try with --fix-missing?
```

## Expected results

The job containers should have the same MTU as the worker container. Alternatively, concourse should offer an option to configure the MTU for containerd.

## Actual results

First, look at the result from `docker exec`ing into the concourse container:

```console
$ docker exec -it concourse_worker_1 bash           
root@6ce4541bb038:/src# ip link show eth0
26: eth0@if27: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1480 qdisc noqueue state UP group default 
    link/ether 02:42:ac:13:00:04 brd ff:ff:ff:ff:ff:ff link-netnsid 0
    inet 172.19.0.4/16 brd 172.19.255.255 scope global eth0
       valid_lft forever preferred_lft forever
```

Note the interface has `mtu 1480`.

Now, hijack the container from the build:

```console
$ fly -t dev hijack -j example/mtu-example
root@9f7acdff-01d8-4acf-75b3-a7864737234c:/tmp/build/c3499c27# ip link show eth0
3: eth0@if13: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP group default 
    link/ether c6:e4:dc:fe:d2:63 brd ff:ff:ff:ff:ff:ff link-netnsid 0
    inet 10.80.0.12/16 brd 10.80.255.255 scope global eth0
       valid_lft forever preferred_lft forever
    inet6 fe80::c4e4:dcff:fefe:d263/64 scope link 
       valid_lft forever preferred_lft forever
```

Note the interface has `mtu 1500`.

## Additional context

My production cluster's containers have an MTU of 1480 because of the tunneling used by the kubernetes CNI plugin I am using. I suspect many kubernetes clusters running concourse will run into this hard to debug issue.

## Triaging info

* Concourse version: 7.0.0
* Browser (if applicable): N/A
* Did this used to work? As far as I know, no

### PR DESCRIPTION
**Pull request by aoldershaw**
[imperative mood]: https://chris.beams.io/posts/git-commit/#imperative

## What does this PR accomplish?

**Bug Fix** | Feature | Documentation

closes #6613
potentially fixes #6599 

## Changes proposed by this PR:

* Match Guardian's behaviour of inferring the MTU to use for the container's bridge network by finding the network interface corresponding to the host's external IP - see https://github.com/cloudfoundry/guardian/blob/15a56cca9deaa1ad04c9c10f91054e7ce82bf3c0/guardiancmd/command.go#L420-L426
  * As with Guardian, the MTU and external IP can still be configured

## Release Note

* In prior versions of Concourse, the Containerd runtime always set the MTU of the container bridge network to the system default
* Now, the Containerd matches Guardian's behavior by:
  * Detecting the external IP of the host (can be set explicitly using `CONCOURSE_CONTAINERD_EXTERNAL_IP`)
  * Extracting the MTU from the network interface corresponding with that IP (can be set explicitly using `CONCOURSE_CONTAINERD_MTU`)

## Contributor Checklist

- [ ] Followed [Code of conduct], [Contributing Guide] & avoided [Anti-patterns]
- [ ] [Signed] all commits
- [ ] Added tests (Unit and/or Integration)
- [ ] Updated [Documentation]
- [ ] Added release note (Optional)

[Code of Conduct]: https://github.com/concourse/concourse/blob/master/CODE_OF_CONDUCT.md
[Contributing Guide]: https://github.com/concourse/concourse/blob/master/CONTRIBUTING.md
[Anti-patterns]: https://github.com/concourse/concourse/wiki/Anti-Patterns
[Signed]: https://help.github.com/en/github/authenticating-to-github/signing-commits
[Documentation]: https://github.com/concourse/docs

## Reviewer Checklist

- [ ] Code reviewed
- [ ] Tests reviewed
- [ ] Documentation reviewed
- [ ] Release notes reviewed
- [ ] PR acceptance performed
- [ ] New config flags added? Ensure that they are added to the
  [BOSH](https://github.com/concourse/concourse-bosh-release) and
  [Helm](https://github.com/concourse/helm) packaging; otherwise, ignored for
  the [integration
  tests](https://github.com/concourse/ci/tree/master/tasks/scripts/check-distribution-env)
  (for example, if they are Garden configs that are not displayed in the
  `--help` text).

### CODE REVIEW

File: worker/workercmd/worker_linux.go [hunk 1]
```diff
@@ -43,12 +43,17 @@ type ContainerdRuntime struct {
...
-	NetworkPool        string    `long:"network-pool" default:"10.80.0.0/16" description:"Network range to use for dynamically allocated container subnets."`
+	Network struct {
+		ExternalIP flag.IP `long:"external-ip" description:"IP address to use to reach container's mapped ports. Autodetected if not specified."`
+		//TODO can DNSConfig be simplifed to just a bool rather than struct with a bool?
+		DNS                DNSConfig `group:"DNS Proxy Configuration" namespace:"dns-proxy"`
+		DNSServers         []string  `long:"dns-server" description:"DNS server IP address to use instead of automatically determined servers. Can be specified multiple times."`
+		RestrictedNetworks []string  `long:"restricted-network" description:"Network ranges to which traffic from containers will be restricted. Can be specified multiple times."`
+		Pool               string    `long:"network-pool" default:"10.80.0.0/16" description:"Network range to use for dynamically allocated container subnets."`
+		MTU                int       `long:"mtu" description:"MTU size for container network interfaces. Defaults to the MTU of the interface used for outbound access by the host. Max allowed value is 1500."`
```
Comment (chenbh) [inline worker/workercmd/worker_linux.go:53]: While it's true [ethernet v2](https://en.wikipedia.org/wiki/Ethernet_frame#Ethernet_II) is capped at 1500, I don't see any enforcement of this (unless it's at the CNI level).

File: worker/mtu/mtu.go [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+		}
+	}
+	return net.Interface{}, fmt.Errorf("no interface found for address %s", ipAddr)
+}
+
+func contains(addresses []net.Addr, ip string) bool {
+	for _, address := range addresses {
+		if strings.HasPrefix(address.String(), ip+"/") {
+			return true
+		}
```
Comment (chenbh) [inline worker/mtu/mtu.go:38]: I'm not fully convinced everyone's network interfaces will report a CIDR block starting with their ip, can we [parse the actual CIDR](https://golang.org/pkg/net/#ParseCIDR) and check to see if that block [contains the ip](https://golang.org/pkg/net/#IPNet.Contains) instead?

**Review by chenbh — COMMENTED** [review]
Left some comments about the way we're figuring out the interface/mtu


Otherwise tested by restricting my docker-compose network to `mtu: 500`, then using `ping -M do` to disable fragmentation:
 1. force a packet that's over the container's mtu
```bash
$ ping -M do -s 2000 github.com
PING github.com (140.82.113.4) 2000(2028) bytes of data.
ping: local error: message too long, mtu=1500
```
 2. trick the container into sending a packet that they think is valid (<1500), but actually invalid to the docker network (>500)
 ```bash
$ ping -M do -s 1000 github.com
PING github.com (140.82.113.4) 1000(1028) bytes of data.
From 10.80.0.1 (10.80.0.1) icmp_seq=1 Frag needed and DF set (mtu = 500)
```
3. send a packet that's valid to the docker network (<500)
```bash
$ ping -M do -s 400
PING github.com (140.82.113.4) 400(428) bytes of data.
408 bytes from lb-140-82-113-4-iad.github.com (140.82.113.4): icmp_seq=1 ttl=36 time=29.7 ms
```

**Comment by chenbh** [comment]
Also found out that docker had support for auto-detecting mtu, removed it, added it again, and then removed it for a second time (https://github.com/moby/moby/pull/18108)

File: worker/mtu/mtu.go (hunk 2 above)
Comment (aoldershaw) [reply worker/mtu/mtu.go]: fair enough - I just copied the logic from guardian so didn't put too much thought into it, but this seems more robust

File: worker/workercmd/worker_linux.go (hunk 1 above)
Comment (aoldershaw) [reply worker/workercmd/worker_linux.go]: also just copied the help text from Guardian - removed the note

**Review by chenbh — APPROVED** [review]
lgtm

**Pull request merged**